File: app/src/main/java/juego/chinchon/Mazo.java
package juego.chinchon;

import android.content.res.Resources;
import android.widget.ImageView;
import java.util.ArrayList;
import java.util.Random;

/**
 * Clase para representar un Mazo de cartas españolas, de 50 cartas. 12 cartas
 * para los cuatro palos más dos comodines.
 * También se utiliza para representar al pozo de descarte, al cual se le
 * colocan las cartas que descarta cada jugador durante el transcurso del juego.
 * @author LeoAM
 */
public class Mazo {
    private static final int MAXCartas = 50;
    private final ArrayList<Carta> cartas;
    private int cantidad;

    /**
     * Constructor de la clase Mazo. Puede crear el mazo vacío o con las cartas.
     * @param vacio Indica si hay que construir el mazo vacío (para el pozo de
     * descarte) o lleno (para el mazo principal).
     */
    public Mazo(boolean vacio) {
        cartas = new ArrayList<>();
        if (vacio) {
            cantidad = 0;
        } else {
            cantidad = MAXCartas;
            setCartas();
            mezclar();
        }
    }

    // Método privado para crear las cartas para el mazo.
    private void setCartas() {
        for (Palo p : Palo.values()) {
            if (p == Palo.Comodin) {
                break;
            }
            for (int j = 1; j <= 12; j++) {
                cartas.add(new Carta(j, p));
            }
        }
        cartas.add(new Carta(0, Palo.Comodin));
        cartas.add(new Carta(0, Palo.Comodin));
    }

    // Método privado para mezclar el mazo.
    private void mezclar() {
        if (cantidad > 1) {
            Random rdm = new Random();
            Carta tmp;
            int j;
            for (int i = cantidad - 1; i > 0; i--) {
                j = rdm.nextInt(i);
                tmp = cartas.get(i);
                cartas.set(i, cartas.get(j));
                cartas.set(j, tmp);
            }
        }
    }

    /**
     * Roba una carta del mazo.
     * @return La carta robada.
     */
    public Carta robar() {
        cantidad--;
        return cartas.remove(0);
    }

    /**
     * Muestra cuál es la carta en el tope del mazo. No la quita.
     * @return La carta en el tope.
     */
    public Carta tope() {
        if(cartas.isEmpty()) {
            return null;
        } else {
            return cartas.get(0);
        }
    }

    /**
     * Coloca una carta al tope del mazo.
     * @param c La carta a colocar.
     */
    public void colocar(Carta c) {
        if (cantidad < MAXCartas) {
            cartas.add(0, c);
            cantidad++;
        }
    }

    /**
     * Vuelca las cartas de un mazo en éste. Luego las mezcla.
     * @param m El mazo de donde se sacan las cartas.
     */
    public void volcar(Mazo m) {
        this.cantidad = m.cantidad;
        m.cantidad = 0;
        for (int i = 0; i < this.cantidad; i++) {
            this.cartas.add(m.cartas.remove(0));
        }
        this.mezclar();
    }

    /**
     * Reparte las cartas entre los jugadores. Reparte 8 cartas para el primer
     * jugador y 7 para el segundo.
     * @param jugadores Lista de jugadores.
     */
    public void repartir(ArrayList<Jugador> jugadores) {
        // Se reparte solo al inicio de la partida.
        if (cantidad == MAXCartas) {
            int cantJugadores = jugadores.size(), j;

            // Vacío las manos de los jugadores:
            for (Jugador jug : jugadores) {
                jug.vaciarMano();
            }

            // Ciclo de cartas
            for (int i = 0; i < 7; i++) {
                for (j = 0; j < cantJugadores; j++) {
                    jugadores.get(j).getMano().addCarta(this.robar());
                }
            }

            // Carta extra para el primer jugador de la ronda:
            jugadores.get(0).getMano().addCarta(this.robar());
        }
    }

    @Override
    public String toString() {
        String str = "";
        for (int i = 0; i < cantidad; i++) {
            str = str.concat(cartas.get(i).toString() + "\n");
        }
        return str;
    }

    /**
     * Devuelve la cantidad de cartas en el mazo.
     * @return Devuelve la cantidad de cartas en el mazo.
     */
    public int getCantidad() {
        return cantidad;
    }

    /**
     * Dada un componente de imagen, le coloca la imagen del tope del mazo.
     * @param iv Componente de ImageView.
     * @param oculto Si se trata del mazo de robo, entonces muestra el dorso de
     * la carta en lugar de la imagén de la carta en particular.
     */
    public void setImagenTope(ImageView iv, boolean oculto) {
        Resources res = iv.getResources();
        if (cartas.isEmpty()) {
            /* Si el mazo o la pila están vacíos, entonces dejó vacío el espacio
            para la imagen. */
            iv.setImageResource(res.getIdentifier("vacio", "drawable", "juego.chinchon"));
        } else
        if (oculto) {
            iv.setImageResource(res.getIdentifier("dorso", "drawable", "juego.chinchon"));
        } else {
            iv.setImageResource(res.getIdentifier(cartas.get(0).getImagePath(), "drawable", "juego.chinchon"));
        }
    }

    /**
     * Comprueba que el mazó esté vacío.
     * @return Si el mazo está vacío o no.
     */
    public boolean vacio() {
        return cartas.isEmpty();
    }
}

File: app/src/main/java/juego/chinchon/Mazo.kt
package juego.chinchon

import android.os.Parcel
import android.os.Parcelable
import java.util.*

/**
 * Clase para representar un Mazo de cartas españolas, de 50 cartas. 12 cartas
 * para los cuatro palos más dos comodines.
 * También se utiliza para representar al pozo de descarte, al cual se le
 * colocan las cartas que descarta cada jugador durante el transcurso del juego.
 *
 * @param vacio Indica si hay que construir el mazo vacío (para el pozo de descarte) o lleno (para el mazo principal).
 * @author LeoAM
 */
class Mazo(vacio: Boolean): Parcelable {

    private var cartas: ArrayList<Carta> = ArrayList()

    /**
     * Devuelve la cantidad de cartas en el mazo.
     * @return Devuelve la cantidad de cartas en el mazo.
     */
    private val cantidad: Int
        get() = cartas.size

    @Suppress("UNCHECKED_CAST")
    constructor(parcel: Parcel) : this(true) {
        cartas = parcel.readArrayList(Carta::class.java.classLoader) as ArrayList<Carta>
    }

    init {
        if (!vacio) {
            setCartas()
            mezclar()
        }
    }

    /**
     * Crea la baraja de cartas y las inserta en este mazo. La baraja consiste
     * en las 12 cartas de los 4 palos y los dos comodines.
     */
    private fun setCartas() {
        for (p in Palo.values()) {
            if (p == Palo.Comodin) {
                break
            }
            for (j in 1..12) {
                cartas.add(Carta(j, p))
            }
        }
        cartas.add(Carta(25, Palo.Comodin))
        cartas.add(Carta(25, Palo.Comodin))
    }

    /**
     * Mezcla las cartas del mazo.
     */
    private fun mezclar() {
        if (cantidad > 2) {
            val rdm = Random()
            var tmp: Carta
            var j: Int
            for (i in cantidad - 1 downTo 1) {
                j = rdm.nextInt(i)
                tmp = cartas[i]
                cartas[i] = cartas[j]
                cartas[j] = tmp
            }
        }
    }

    /**
     * Roba una carta del mazo.
     *
     * @return La carta robada.
     */
    fun robar(): Carta {
        return cartas.removeAt(0)
    }

    /**
     * Muestra cuál es la carta en el tope del mazo. No la quita.
     *
     * @return La carta en el tope.
     */
    fun tope(): Carta {
        return cartas[0]
    }

    /**
     * Coloca una carta en el tope del mazo.
     *
     * @param c La carta a colocar.
     */
    fun colocar(c: Carta) {
        if (cantidad < MAXCartas) {
            cartas.add(0, c)
        }
    }

    /**
     * Vuelca las cartas de un mazo en éste. Luego las mezcla.
     *
     * @param m El mazo de donde se sacan las cartas.
     */
    fun volcar(m: Mazo) {
        for (i in 0 until cantidad) {
            val carta = m.cartas.removeAt(0)
            cartas.add(carta)
        }
        mezclar()
    }

    /**
     * Reparte las cartas entre los jugadores. Reparte 8 cartas para el primer
     * jugador y 7 para el segundo.
     *
     * @param jugadores Lista de jugadores.
     */
    fun repartir(jugadores: ArrayList<Jugador>) {
        if (cantidad == MAXCartas) {
            for (jugador in jugadores) {
                jugador.vaciarMano()
            }

            for (i in 0 until 7) {
                for (j in 0 until jugadores.size) {
                    jugadores[j].mano.addCarta(robar())
                }
            }
        }
    }

    /**
     * Devuelve la representación en String de este mazo como una lista de
     * todas las cartas actualmente en el mazo, cada una en una línea.
     *
     * @return Representación en String de este mazo.
     */
    override fun toString(): String {
        var str = ""
        for (i in 0 until cantidad) {
            str = str + cartas[i].toString() + "\n"
        }
        return str
    }

    /**
     * Comprueba que el mazó esté vacío.
     *
     * @return Si el mazo está vacío o no.
     */
    fun vacio(): Boolean {
        return cartas.isEmpty()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeList(cartas as List<*>?)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Mazo> {
        override fun createFromParcel(parcel: Parcel): Mazo {
            return Mazo(parcel)
        }

        override fun newArray(size: Int): Array<Mazo?> {
            return arrayOfNulls(size)
        }

        private const val MAXCartas = 50
    }
}
